feat: add-item-requests & tests

dmitry-razumov opened the pull request:
feat: add-item-requests
feat: tests

dmitry-razumov committed:
feat: add-item-requests & tests

TarasovSam commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -48,16 +53,22 @@ public BookingResponseDto getById(@RequestHeader("X-Sharer-User-Id") long userId
-        return mapper.toResponseDto(service.getAllByUser(userId, state));
+                                                 @RequestParam(defaultValue = "ALL") String state,
+                                                 @RequestParam(defaultValue = "0") @PositiveOrZero int from,
+                                                 @RequestParam(defaultValue = "20") @Positive int size) {
```
Хорошо, используешь для проверки данных параметров штатные инструменты валидации в виде аннотации

TarasovSam commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -24,9 +25,11 @@ public class Booking {
     private LocalDateTime start;
     @Column(name = "end_date")
     private LocalDateTime end;
```
Время начала и время окончания не валидируются?

dmitry-razumov replied:
fix

TarasovSam commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -72,7 +77,7 @@ List<Booking> findAllByOwnerIdAndStartAfterOrderByStartDesc(
             "and b.status = ?2 " +
```
Можно отойти от абстрактных чисел в запросе к именованным параметрам)
https://www.baeldung.com/spring-data-jpa-query
(Раздел Named Parameters)
Аналогично в других местах

dmitry-razumov replied:
fix

TarasovSam commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -95,78 +97,80 @@ public Booking getById(long bookingId, long userId) {
+    public List<Booking> getAllByUser(long userId, String state, int from, int size) {
         User booker = findUserByIdOrThrow(userId);
         LocalDateTime currentTime = LocalDateTime.now();
+        Pageable page = PageRequest.of(from / size, size);
```
Рекомендую сделать настройки пагинации немного проще:
Сформируй класс, наследуемый от PageRequest и определи для него конструктор, принимающий параметры from и size. Данный класс будет реализовывать логику расчёта пагинации. Таким образом алгоритм расчёта пагинации будет инкапсулирован в данный класс.

Это позволит расширить стандартный spring класс пагинации под твою бизнес логику

dmitry-razumov replied:
попробовал реализовать

TarasovSam commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -20,12 +19,16 @@ public class Item {
     private String name;
+    @NotBlank
     private String description;
     private Boolean available;
```
Вроде бы available не может быть пустым

dmitry-razumov replied:
fix

TarasovSam commented on src/main/java/ru/practicum/shareit/request/ItemRequestController.java:
```diff
@@ -1,12 +1,59 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    ItemRequestDto create(@RequestHeader(name = "X-Sharer-User-Id") long userId,
```
Модификаторы эндпоинтов почему дефолтные?

dmitry-razumov replied:
fix

TarasovSam commented on src/main/java/ru/practicum/shareit/request/dto/ItemInRequestField.java:
```diff
@@ -0,0 +1,14 @@
+
+@Data
+@Builder
+public class ItemInRequestField {
```
Валидация?

dmitry-razumov replied:
поля ItemInRequestField только в дто ответа
добавил аннотацию mapStruct'у игнорировать при маппинге из дто при приеме в 
@Mapping(target = "items", ignore = true)
 ItemRequest toItemRequest(ItemRequestDto itemRequestDto);

TarasovSam commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -16,7 +18,9 @@ public class User {
     private String name;
+    @Email
     @Column(unique = true)
     private String email;
```
Почта же не может быть пустой

dmitry-razumov replied:
fix

TarasovSam commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Test
+    @SneakyThrows
```
Да, в принципе аннотация удобна, но ее следует аккуратно использовать, т.к. она грубо говоря обезличивает
https://urvanov.ru/2015/09/20/lombok-sneakythrows-%D1%85%D1%80%D0%B0%D0%B1%D1%80%D0%BE-%D0%B1%D1%80%D0%BE%D1%81%D0%B0%D0%B5%D0%BC-%D0%BF%D1%80%D0%BE%D0%B2%D0%B5%D1%80%D1%8F%D0%B5%D0%BC%D1%8B%D0%B5-%D0%B8%D1%81%D0%BA/
Прочитай этот материал, там более подробно расписано, когда ее можно использовать)
Я бы писал вручную

dmitry-razumov replied:
ок. спасибо

TarasovSam commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+                .build();
+    }
+
+    static Booking createBooking() {
```
Вместо статического блока создания элементов, это можно выделить в @beforeeach. Что гарантирует начало каждого теста с "чистого листа"
В других тестах так же

dmitry-razumov replied:
fix

TarasovSam commented on src/test/java/ru/practicum/shareit/booking/BookingResponseDTOTest.java:
```diff
@@ -0,0 +1,89 @@
+        assertThat(result).extractingJsonPathStringValue("$.booker.name").isEqualTo("booker");
+        assertThat(result).extractingJsonPathStringValue("$.booker.email")
+                .isEqualTo("booker@mail.com");
+        assertThat(result).extractingJsonPathStringValue("$.status").isEqualTo("WAITING");
```
Здесь ты делаешь множественные утверждения (assertThat). Это не всегда плохо, но может затруднить определение, что именно пошло не так, если тест не пройдет.
Думаю, стоит разделить эти тесты на более мелкие. И тогда чтобы не создавать для каждого booker, owner и т.д., нужно выделить это в @beforeeach

dmitry-razumov replied:
fix

TarasovSam reviewed (commented):
Дмитрий, привет!
Проверил твое задание, все довольно хорошо. Есть что скорректировать и что улучшить, оставил соответствующие комментарии)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

dmitry-razumov committed:
fix: валидация
fix: named param query
add: custom PageRequest
fix: модификаторы эндпоинтов
fix: to @beforeEach
refact: json test

TarasovSam commented on src/main/java/ru/practicum/shareit/utils/PageRequestCustom.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+
+public class PageRequestCustom extends PageRequest {
```
Хорошо)

TarasovSam commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -70,11 +72,15 @@ static Booking createBooking() {
-    void shouldCreateBooking() {
-        Booking booking = createBooking();
-        BookingRequestDto requestDto = createBookingRequest();
+    void shouldCreateBooking() throws Exception {
```
Здесь ты пробрасываешь исключение дальше, а следует обработать в методе
Аналогично и дальше

TarasovSam reviewed (commented):
Дмитрий, проверил твое задание, спасибо за работу и за то, что дорабатываешь код. Оставил дополнительные комментарии)
Задание я принимаю, успеха в обучении и до следующего спринта!) Если будут вопросы, пиши в пачку)
